src/FSharpVSPowerTools.Core/Utils.fs

```diff
@@ -89,32 +186,94 @@ module Array =
         checkNonNull "array" array
         let temp = Array.zeroCreate array.Length
         let mutable i = 0
-
-        let hashSet = HashSet<'T>(HashIdentity.Structural<'T>)
+        let hashSet = HashSet<'T> HashIdentity.Structural<'T>
         for v in array do 
             if hashSet.Add(v) then
                 temp.[i] <- v
-                i <- i + 1
+                &i += 1
+        temp.[0..i-1]
 
-        Array.sub temp 0 i 
 
-    /// Optimized arrays equality. ~100x faster than `array1 = array2`.
-    let inline areEqual (x: 'a[]) (y: 'a[]) =
-        match x, y with
-        | null, null -> true
-        | [||], [||] -> true
-        | null, _ | _, null -> false
-        | _ when x.Length <> y.Length -> false
-        | _ ->
-            let mutable break' = false
-            let mutable i = 0
-            let mutable result = true
-            while i < x.Length && not break' do
-                if x.[i] <> y.[i] then 
-                    break' <- true
-                    result <- false
-                i <- i + 1
-            result
+    let distinctBy keyf (array:'T[]) =
+        checkNonNull "array" array
+        let temp = Array.zeroCreate array.Length
+        let mutable i = 0 
+        let hashSet = HashSet<_> HashIdentity.Structural<_>
+        for v in array do
+            if hashSet.Add(keyf v) then
+                temp.[i] <- v
+                &i += 1
+        temp.[0..i]
```

Comment:
I used the FSharp.Core implementation
https://github.com/Microsoft/visualfsharp/blob/fsharp4/src/fsharp/FSharp.Core/array.fs#L278-L287